### PR TITLE
Hipify revamp changes for apex extensions on ROCm.

### PR DESCRIPTION
**Pull request by lcskrishna**
Hipify changes based on upstream pytorch PR : https://github.com/pytorch/pytorch/pull/48715
Had changes for backward compatibility when the torch version >= 1.5 && < 1.8 

cc: @jithunnair-amd @jeffdaily @sunway513

### CODE REVIEW

**Comment by lcskrishna** [comment]
jenkins: retest this please

**Comment by jeffdaily** [comment]
I thought the point of the hipify revamp was to not need to call hipify explicitly any longer?  The source lists for both before and after the hipify revamp change contain "hip" in their names -- I thought CUDAExtension was auto-hipifying as needed?  Is that not working?

Having to support backward-compat in this PR is quite inelegant.  Ugly really.  Wondering if we can't just rip off the bandaid and be done with it.  Move forward with the hipify revamp.  We have complete control over our APEX fork, we aren't trying to upstream.  How strong is our need for backwards-compat here?

**Comment by sunway513** [comment]
We're okay to drop backward-compatibility for APEX on the master branch if needed; however, we do need to document this change in README.

**Comment by lcskrishna** [comment]
Sure, I'm okay to drop the backward compatibility. The ugliness is created due to the backward compatibility issues. If any of the internal teams doesn't want to use any older versions, I'm ready to drop it off.

**Comment by lcskrishna** [comment]
@jeffdaily @sunway513 Ready for review based on latest hipify changes. Deprecated the older style, and added a note in ReadMe.

**Review by jithunnair-amd — APPROVED** [review]
Looks good to me!

File: README.md [hunk 1]
```diff
@@ -119,6 +119,7 @@ See the [Docker example folder](https://github.com/NVIDIA/apex/tree/master/examp
 * Python 3.6
 * Pytorch 1.5 or newer, The HIPExtensions require 1.5 or newer.
 * We recommend follow the instructions from [ROCm-Pytorch](https://github.com/ROCmSoftwarePlatform/pytorch) to install pytorch on ROCm.
+* Note: For Pytorch versions less than 1.8 please consider release package [ROCm-Apex v0.3](https://github.com/ROCmSoftwarePlatform/apex/releases/tag/v0.3) . For pytorch versions >= 1.8, use the master branch as this contains latest hipify revamp changes.
```
Comment (jithunnair-amd) [inline README.md:122]: Should we reword this to give a stronger guidance that building ROCm apex from source is *not supported* anymore for pytorch < 1.8? That's my understanding, please correct me if I'm wrong.

File: setup.py [hunk 2]
```diff
@@ -167,77 +161,32 @@ def check_if_rocm_pytorch():
...
-                                       'csrc/hip/multi_tensor_adagrad.hip',
-                                       'csrc/hip/multi_tensor_novograd.hip',
-                                       'csrc/hip/multi_tensor_lamb.hip'],
-                              extra_compile_args=['-O3'] + version_dependent_macros))
+                              extra_compile_args = { 'cxx' : ['-O3'] + version_dependent_macros,
+                                                     'nvcc': nvcc_args_multi_tensor if not is_rocm_pytorch else hipcc_args_multi_tensor}))
 
-        if not is_rocm_pytorch:
-            ext_modules.append(
+        print ("INFO: Builidng syncbn extension.")
```
Comment (jeffdaily) [inline setup.py:167]:
```suggestion
        print ("INFO: Building syncbn extension.")
```

Comment (lcskrishna) [reply setup.py]: Done

File: README.md (hunk 1 above)
Comment (lcskrishna) [reply README.md:122]: We can do that. I would let @sunway513 or @jeffdaily to comment on this one.

**Comment by jithunnair-amd** [comment]
Nit: we can get rid of `from torch.utils.hipify import hipify_python` in `setup.py`.

File: README.md (hunk 1 above)
Comment (jeffdaily) [reply README.md:122]: Keep the suggestion to use a release package for pytorch versions < 1.8.  Add the stronger guidance that building rocm apex from source is not supported for pytorch versions < 1.8.

**Comment by jeffdaily** [comment]
> Nit: we can get rid of `from torch.utils.hipify import hipify_python` in `setup.py`.

@lcskrishna please do this, as well.

File: README.md (hunk 1 above)
Comment (jithunnair-amd) [reply README.md:122]: Thanks @jeffdaily. @lcskrishna Once you make this change and the other nit change, we should merge this in. I'll need it to get the DeepSpeed CI back on track.

Comment (lcskrishna) [reply README.md]: Done

**Comment by lcskrishna** [comment]
@jeffdaily @jithunnair-amd updated as per the comments before.
